### ISSUE TITLE
Server instance failing to use TLS when hostname and SANs are present in PKI

### ISSUE THREAD

**Issue by RedSofaForEveryone** [hashicorp/docker-vault#143]
I have generated my certs from Pfsense. These certs have worked for all other servers except vault.

docker-compose.yml
```
version: "3"
services:
  vault:
    image: vault:latest
    container_name: vault01
    hostname: vault01
    extra_hosts:
      - "vault01:127.0.0.1"
      - "vault01.lab.com:127.0.0.1"
      - "vault01.lab.lab:127.0.0.1"
    ports:
      - "8200:8200"
    volumes:
      - "./data:/vault/data"
      - "./certs:/vault/certs"
      - "./hostname:/etc/hostname"
    cap_add:
      - IPC_LOCK
    environment:
      VAULT_LOCAL_CONFIG: |
        ui = true
        default_lease_ttl = "768h"
        storage "file" {
          path = "/vault/data"
        }

        listener "tcp" {
          address = "127.0.0.1:8200"
          tls_cert_file = "/vault/certs/vault01.lab.com.bundle.crt"
          tls_key_file = "/vault/certs/vault01.lab.com.key"
          tls_client_ca_file = "/vault/certs/lab-servers-ca.crt"
          tls_require_and_verify_client_cert = "true"
        }
    command: server

    stdin_open: "true"
    tty: "true"
```

`vault01.lab.com.bundle.crt` contains the public cert and the CA afterwards bundled in.

The common name is `vault01`, with 3 SAN's:

```
DNS=vault01
DNS=vault01.lab.com
DNS=vault01.lab.lab
```

I added `vault01.lab.com` to my local `/etc/hosts`

Trying to connect via the web browser via `vault01.lab.com:8200` tells me:

```
The page you are trying to view cannot be shown because the authenticity of the received data could not be verified.
```

Doing a `docker exec -it vault01 vault init` gives me:
```
Error initializing: Put https://127.0.0.1:8200/v1/sys/init: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
```

I `cat`'d `$HOSTNAME` and it gives me `vault01` thanks to the docker-compose.yml.

I thought it may be the container's `/etc/hosts` file, so I mounted `/etc/hosts` from the current directory and it still gives me the error.

Why is this not working?

Thank you.

Also, just to confirm that the certs are set up correctly, this is the output from sslshopper's cert decoder:

```
Common Name: vault01
Subject Alternative Names: vault01, vault01.lab.com, vault01.lab.lab
Organization: lab
Organization Unit: vault
Locality: xxx
State: xx
Country: xx
Valid From: April 4, 2019
Valid To: April 1, 2029
Issuer: lab-servers-ca, lab
Serial Number: 21 (0x15)
```

**Comment by Constantin07** [comment]
@RedSofaForEveryone looks like you need to add `127.0.0.1` to the certificate request SAN IPs.

**Comment by RedSofaForEveryone** [comment]
@Constantin07 but why should I have to do that if the hostname is `vault01` and the cert points to `vault01`, as well as the SAN names all pointing to `vault01...`?

That config should be accurate, because if that was the case, I would have to make a new SAN IP for everytime this machine was moved in some way.

**Comment by RedSofaForEveryone** [comment]
So the answer was make your `address` set to `0.0.0.0:<port>`.

**Comment by RedSofaForEveryone** [comment]
Actually, there is another issue:

I had to set the environment variable `VAULT_TLS_SERVER_NAME` in the docker compose file to `vault01.lab.com`.

^This stopped giving me the IP SAN error locally, but I still get x509 unknown certificate authority error.

I have confirmed the browser and using curl remotely can both get responses through without https errors.

To fix the `docker exec` local commands, I had to manually place the CA inside the containers `/etc/ssl/certs/` folder.
